strawberryfields/backends/states.py

```diff
@@ -1405,3 +1415,412 @@ def all_fock_probs(self, **kwargs):
+        super().__init__(num_modes, mode_names)
+
+        self._data = state_data
+        self.num_weights = num_weights
+        # vector of means and covariance matrix, using frontend x,p scaling
+        self._mus = self._data[0] * np.sqrt(self._hbar / 2)
+        self._covs = self._data[1] * (self._hbar / 2)
+        self._weights = self._data[2]
+
+        self._basis = "bosonic"
+        self._str = "<BosonicState: num_modes={}, num_weights={}, pure={}, hbar={}>".format(
+            self.num_modes, self.num_weights, self._pure, self._hbar
+        )
+
+    def __eq__(self, other):
+        """Equality operator for BaseBosonicState.
+
+        Returns True if other BaseBosonicState is close to self.
+        This is done by comparing the weights, means vectors and covs matrices.
+        If both are within the EQ_TOLERANCE, True is returned.
+
+        Args:
+            other (BaseGaussianState): BaseGaussianState to compare against.
+        """
+        # pylint: disable=protected-access
+        # TODO: check equality for two different representations of the same state.
+        # This only checks if they have equal representations.
+        if not isinstance(other, type(self)):
+            return False
+
+        if self.num_modes != other.num_modes:
+            return False
+
+        if self.num_weights != other.num_weights:
+            return False
+
+        if (
+            np.allclose(self._mus, other._mus, atol=self.EQ_TOLERANCE, rtol=0)
+            and np.allclose(self._covs, other._covs, atol=self.EQ_TOLERANCE, rtol=0)
+            and np.allclose(self._weights, other._weights, atol=self.EQ_TOLERANCE, rtol=0)
+        ):
+            return True
+
+        return False
+
+    def means(self):
+        r"""The vectors of means describing the Bosonic state.
+
+        Returns:
+          array: a num_weights by :math:`2N` array.
+        """
+        return self._mus
+
+    def covs(self):
+        r"""The covariance matrices describing the Bosonic state.
+
+        Returns:
+          array: a num_weights by :math:`2N\times 2N` array.
+        """
+        return self._covs
+
+    def weights(self):
+        r"""The weights describing the Bosonic state.
+
+        Returns:
+          array: an array of length num_weights.
+        """
+        return self._weights
+
+    def purity(self):
+        r"""Returns the purity of the state."""
+
+        pur = 0
+        for i, weight_i in enumerate(self._weights):
+            exp_arg = np.einsum(
+                "...j,...jk,...k",
+                (self._mus[i] - self._mus),
+                np.linalg.inv(self._covs + self._covs[i]),
+                (self._mus[i] - self._mus),
+            )
+            prefactor = 1 / (np.sqrt(np.linalg.det((self._covs + self._covs[i]))))
+            pur += np.sum((self._weights * weight_i * prefactor) * np.exp(-0.5 * exp_arg))
+        pur *= sf.hbar ** self.num_modes
```

Comment:
Should this be `self._hbar`? It might be important to use the value of hbar set at state creation, rather than the global hbar value.

Because the global hbar value might change _after_ this state was created!

Reply:
```suggestion
        pur *= self._hbar ** self.num_modes
```